### PR TITLE
Support function array/broadcast_to

### PR DESCRIPTION
**Pull request by youchan**
Support function array/broadcast_to

## example

```ruby
> require 'chainer'
> x = Numo::SFloat.new(3).seq
> h = Chainer::Functions::Array::BroadcastTo.broadcast_to(x, [3,2,3])
> h.data
=> Numo::SFloat#shape=[3,2,3]
[[[0, 1, 2],
  [0, 1, 2]],
 [[0, 1, 2],
  [0, 1, 2]],
 [[0, 1, 2],
  [0, 1, 2]]]
```

### CODE REVIEW

File: lib/chainer/functions/array/broadcast_to.rb [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+      class BroadcastTo < Function
+        def initialize(shape)
+            @shape = shape
+        end
+
+        def self.broadcast_to(xs, shape)
+          self.new(shape).(xs)
+        end
+
+        def forward_cpu(xs)
```
Comment (naitoh) [inline lib/chainer/functions/array/broadcast_to.rb]: Here, you define the forward_cpu(), backward_cpu() method.
If there is no plans to separate the code between the CPU and the GPU in the future, I think that the forward(), backward() names are better as in the original chainer/functions/array/broadcast.py.

Comment (youchan) [reply lib/chainer/functions/array/broadcast_to.rb]: @naitoh fix it.

File: lib/chainer/functions/array/broadcast_to.rb [hunk 2]
```diff
@@ -0,0 +1,43 @@
+module Chainer
+  module Functions
+    module Array
+      # Function that broadcasts an array to a new shape."""
```
Comment (hatappi) [inline lib/chainer/functions/array/broadcast_to.rb]: I think `"""` is how to write comments when using Python

File: lib/chainer/functions/array/broadcast_to.rb [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+      class BroadcastTo < Function
+        def initialize(shape)
+            @shape = shape
+        end
+
+        def self.broadcast_to(xs, shape)
+          self.new(shape).(xs)
+        end
+
+        def forward(xs)
```
Comment (hatappi) [inline lib/chainer/functions/array/broadcast_to.rb]: When using the `forward` function it is necessary to initialize with `retain_inputs`
ref. https://github.com/red-data-tools/red-chainer/blob/master/lib/chainer/functions/activation/relu.rb#L30

Comment (youchan) [reply lib/chainer/functions/array/broadcast_to.rb]: Why is there such an implicit rule?
It should be in caller of `foward` if always need it.

**Comment by youchan** [comment]
@hatappi I fixed https://github.com/red-data-tools/red-chainer/pull/60#discussion_r206909139 https://github.com/red-data-tools/red-chainer/pull/60#discussion_r206911797

File: lib/chainer/functions/array/broadcast_to.rb (hunk 3 above)
Comment (naitoh) [reply lib/chainer/functions/array/broadcast_to.rb]: Please See below. 

https://github.com/chainer/chainer/blob/v2.0.2/chainer/function.py#L127
or
https://docs.chainer.org/en/v2.0.2/reference/core/generated/chainer.Function.html#chainer.Function

And, I think that it is better to reflect this content in the comment of lib/chainer/function.rb.

Comment (youchan) [reply lib/chainer/functions/array/broadcast_to.rb]: Thank you! I see.
I suggest it is not necessary to call `retain_input` when inputs is not used in backward. We should not inherit bad interface from chainer.

Comment (naitoh) [reply lib/chainer/functions/array/broadcast_to.rb]: With the change to Chainer 3.0 ("the new-style differentiable function") of Red-Chainer base code in the future, I think that initialization processing of retain_inputs becomes unnecessary.

Please read the following.
https://files.gitter.im/red-data-tools/ja/IN8R/IMG_5654.JPG
https://github.com/chainer/chainer/pull/2970
https://github.com/chainer/chainer/pull/3420

Comment (hatappi) [reply lib/chainer/functions/array/broadcast_to.rb]: > With the change to Chainer 3.0 ("the new-style differentiable function") of Red-Chainer base > code in the future, I think that initialization processing of retain_inputs becomes unnecessary.

yes.
The base code of red-chainer is currently proceeding to correspond to Chainer 3.0.
Now We are porting Chainer's code as it is, but I'd like to gradually improve it.

**Comment by hatappi** [comment]
Please resolve conflicts 🙏

**Comment by youchan** [comment]
I have rebased. @hatappi

**Comment by hatappi** [comment]
@youchan 
Sorry for the late review.
Thank you for contributing 🎉

**Pull request merged**